### PR TITLE
Transaction pruning service

### PR DESCRIPTION
**Pull request by poornachandra**
https://issues.apache.org/jira/browse/TEPHRA-203
https://issues.apache.org/jira/browse/TEPHRA-199

This PR has three things -
 * Define and enforce maximum transaction lifetime where transactions can be used for data writes
 * Add `TransactionPruningService` to periodically schedule transaction prune run
* Rename package name janitor to txprune

It would be easier to review the commits independently.

### CODE REVIEW

File: tephra-core/src/main/java/org/apache/tephra/TxConstants.java [hunk 1]
```diff
@@ -158,6 +158,14 @@
      * The default value for the transaction timeout limit, in seconds: unlimited.
      */
     public static final int DEFAULT_TX_MAX_TIMEOUT = Integer.MAX_VALUE;
+    /**
+     * The maximum time in seconds that a transaction can be used for data writes.
+     */
+    public static final String CFG_TX_MAX_LIFETIME = "data.tx.max.lifetime";
```
Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TxConstants.java]: This is only for short transactions right? Wasn't very clear from the comment.

Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TxConstants.java]: Nevermind, it looks like it is for all transactions.

File: tephra-core/src/main/java/org/apache/tephra/TxConstants.java [hunk 2]
```diff
@@ -158,6 +158,14 @@
      * The default value for the transaction timeout limit, in seconds: unlimited.
      */
     public static final int DEFAULT_TX_MAX_TIMEOUT = Integer.MAX_VALUE;
+    /**
+     * The maximum time in seconds that a transaction can be used for data writes.
```
Comment (anew) [inline tephra-core/src/main/java/org/apache/tephra/TxConstants.java]: Does this apply to all transaction, including long-running ones? I think it helps to say that after this time, all writes to the table will fail, and that is ensured by Tephra's coprocessors.

You should also document that this value influences invalid tx pruning.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/TransactionAwareHTable.java [hunk 3]
```diff
@@ -649,6 +649,7 @@ private Delete transactionalizeAction(Delete delete) throws IOException {
         txDelete.setAttribute(entry.getKey(), entry.getValue());
     }
     txDelete.setDurability(delete.getDurability());
+    addToOperation(txDelete, tx);
```
Comment (anew) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/TransactionAwareHTable.java]: unrelated change?

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java [hunk 4]
```diff
@@ -179,6 +187,13 @@ public void preGetOp(ObserverContext<RegionCoprocessorEnvironment> e, Get get, L
   }
 
   @Override
+  public void prePut(ObserverContext<RegionCoprocessorEnvironment> e, Put put, WALEdit edit, Durability durability)
+    throws IOException {
+    Transaction tx = getFromOperation(put);
+    ensureValidTxLifetime(tx);
+  }
+
+  @Override
```
Comment (anew) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java]: what about increment operations? and readless increments?

File: tephra-core/src/main/java/org/apache/tephra/TxConstants.java (hunk 2 above)
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/TxConstants.java]: Yes, this limit applies to all transactions. Will update the documentation.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java (hunk 4 above)
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java]: Increment and other operations are non-transactional today. Tephra does not support readless increments.

Comment (anew) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java]: oh i that case it is ok.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/TransactionAwareHTable.java (hunk 3 above)
Comment (anew) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/TransactionAwareHTable.java]: I see

File: tephra-core/src/main/java/org/apache/tephra/txprune/TransactionPruningRunnable.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        LOG.info("Not pruning invalid list since minimum prune upper bound ({}) is less than 1", minPruneUpperBound);
+        return;
+      }
+
+      long[] invalids = tx.getInvalids();
+      TreeSet<Long> toTruncate = new TreeSet<>();
+      LOG.debug("Invalid list: {}", invalids);
+      for (long invalid : invalids) {
+        if (invalid <= minPruneUpperBound) {
+          toTruncate.add(invalid);
```
Comment (anew) [inline tephra-core/src/main/java/org/apache/tephra/txprune/TransactionPruningRunnable.java]: do we have to compute this list? Or can the txManager simply remove all transactions that are older than the minPruneUpperBound?

File: tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java [hunk 6]
```diff
@@ -163,8 +167,22 @@ public static long getPruneUpperBound(Transaction tx) {
...
+    // transaction until it exhausts its max lifetime.
+    long elapsedTime = currentTimeMillis - TxUtils.getTimestamp(maxInvalidTx);
+    long invalidTxBound;
+    if (elapsedTime > txMaxLifetimeMillis) {
+      // maxInvalidTx is past its max lifetime
+      invalidTxBound = maxInvalidTx;
+    } else {
+      // Reduce the lifetime of maxInvalidTx, so that it cannot be used for any more writes
+      long remainingTime = txMaxLifetimeMillis - elapsedTime;
+      invalidTxBound = maxInvalidTx - remainingTime * TxConstants.MAX_TX_PER_MS - 1;
```
Comment (anew) [inline tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java]: Independent of maxInvalidTx, isn't this always the same as (currentTimeMillis - txMaxLifeTimeMillis) * MAX_TX_PER_MS?

Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java]: Good point. Now that I think about it, we can further simplify the method `TxUtils.getPruneUpperBound()` to simply return `(currentTimeMillis - txMaxLifetimeMillis) * TxConstants.MAX_TX_PER_MS - 1`, irrespective of whether there are in-progress transactions or not. Since no transaction (in-progress or invalid) can be used for writes after `txMaxLifetimeMillis`.

What do you say?

File: tephra-core/src/main/java/org/apache/tephra/txprune/TransactionPruningRunnable.java (hunk 5 above)
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/txprune/TransactionPruningRunnable.java]: The reason to build the list is so that we can use the bulk `removeAll()` method of invalid list, rather than removing the invalid ids one by one.

File: tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java (hunk 6 above)
Comment (anew) [inline tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java]: agreed.

Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java]: On further thought, we will need to compute inactive transaction bound separately from prune upper bound. 

Inactive transaction bound can be computed using just current time and tx max lifetime as ` (currentTimeMillis - txMaxLifetimeMillis) * TxConstants.MAX_TX_PER_MS - 1`. 

However, prune upper bound is a record of what invalid data was removed during a major compaction, and hence will need to use the invalid list and in-progress list from the transaction snapshot used at the time of the major compaction.

Comment (anew) [inline tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java]: yes, that comes from each of the regions, right?

Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java]: Yes

**Comment by poornachandra** [comment]
@anew I have separated out the computation of prune upper bound and inactive tx bound as discussed, please take another look.

**Comment by anew** [comment]
LGTM
